### PR TITLE
Fixes from Andy for some BDSTP bugs

### PR DESCRIPTION
**Pull request by bredelings**

### CODE REVIEW

File: src/core/distributions/phylogenetics/tree/birthdeath/BirthDeathSamplingTreatmentProcess.cpp [hunk 1]
```diff
@@ -1590,36 +1594,52 @@ double BirthDeathSamplingTreatmentProcess::simulateDivergenceTime(double origin,
...
     }
 
+    for (size_t i=0; i<global_timeline.size(); ++i)
+    {   
+        size_t idx = global_timeline.size() - i - 1;
+        if ( global_timeline[idx] < t ) {
+            break;
+        } else if (global_timeline[idx] > t)
+        {   
+            // TODO by ignoring time = t we implicitly count all tips at a time as survivors
```
Comment (hoehna) [inline src/core/distributions/phylogenetics/tree/birthdeath/BirthDeathSamplingTreatmentProcess.cpp:1637]: perhaps remove this todo if it's not a todo anymore

File: src/core/distributions/phylogenetics/tree/birthdeath/BirthDeathSamplingTreatmentProcess.cpp [hunk 2]
```diff
@@ -1590,36 +1594,52 @@ double BirthDeathSamplingTreatmentProcess::simulateDivergenceTime(double origin,
...
+        {
+            return 0;
+        }
+        survivors = 2;
+    }
 
-        // my age needs to be smaller that the requested time
-        if ( (my_age - t) < 1E-4 )
+    for (size_t i=0; i<serial_bifurcation_times.size(); ++i) {
+        if (t < serial_bifurcation_times[i])
```
Comment (hoehna) [inline src/core/distributions/phylogenetics/tree/birthdeath/BirthDeathSamplingTreatmentProcess.cpp:1617]: I'm not sure if this checking for the specific time is safe, because our io usually writes branch lengths rounded to 4 decimals.

File: src/core/distributions/phylogenetics/tree/birthdeath/BirthDeathSamplingTreatmentProcess.cpp [hunk 3]
```diff
@@ -1590,36 +1594,52 @@ double BirthDeathSamplingTreatmentProcess::simulateDivergenceTime(double origin,
...
-            {
-                survivors++;
-            }
+            survivors++;
         }
-        else if ( (my_age - t) < -1E-4 )
+    }
+
+    for (size_t i=0; i<serial_tip_ages.size(); ++i) {
+        if (t < serial_tip_ages[i])
```
Comment (hoehna) [inline src/core/distributions/phylogenetics/tree/birthdeath/BirthDeathSamplingTreatmentProcess.cpp:1624]: same as above

File: src/core/distributions/phylogenetics/tree/birthdeath/BirthDeathSamplingTreatmentProcess.cpp [hunk 4]
```diff
@@ -604,7 +612,7 @@ bool BirthDeathSamplingTreatmentProcess::countAllNodes(void) const
               int at_event = whichIntervalTime(t);
 
               // If this bifurcation is not at an event time (and specifically at an event time with Lambda[i] > 0), it's a serial bifurcation
-              if ( at_event == -1 )
+              if ( at_event == -1 || lambda_event[at_event] < DBL_EPSILON)
```
Comment (hoehna) [inline src/core/distributions/phylogenetics/tree/birthdeath/BirthDeathSamplingTreatmentProcess.cpp:615]: why was this removed? I had some tests where the burst events were not seen as burst events because the times didn't exactly match due to rounding. Is this now covered better in at_event?

**Review by hoehna — CHANGES_REQUESTED** [review]
I hope all this works well, but I had a few minor questions which might be good to double check.

File: src/core/distributions/phylogenetics/tree/birthdeath/BirthDeathSamplingTreatmentProcess.cpp (hunk 2 above)
Comment (afmagee) [reply src/core/distributions/phylogenetics/tree/birthdeath/BirthDeathSamplingTreatmentProcess.cpp:1619]: @hoehna We've already addressed the classification-time issue elsewhere though. countAllNodes() is always called first, which uses whichIntervalTime() to classify node ages, and _that_ function uses the 1e-4 tolerance

Comment (afmagee) [reply src/core/distributions/phylogenetics/tree/birthdeath/BirthDeathSamplingTreatmentProcess.cpp:1619]: In other words, we outsource to other functions the matter of "is this (potentially) low-precision tree time equal to this high-precision event-time" and so now we can just compare doubles to doubles, give or take possible a <=

File: src/core/distributions/phylogenetics/tree/birthdeath/BirthDeathSamplingTreatmentProcess.cpp (hunk 1 above)
Comment (afmagee) [reply src/core/distributions/phylogenetics/tree/birthdeath/BirthDeathSamplingTreatmentProcess.cpp:1639]:
```suggestion
            // by ignoring time = t we implicitly count all tips at a time as survivors
```

Yeah, perhaps it's not a to-do anymore but I'd like to leave it as a comment? I don't want to confuse future-me

File: src/core/distributions/phylogenetics/tree/birthdeath/BirthDeathSamplingTreatmentProcess.cpp (hunk 3 above)
Comment (afmagee) [reply src/core/distributions/phylogenetics/tree/birthdeath/BirthDeathSamplingTreatmentProcess.cpp:1626]: ibid

File: src/core/distributions/phylogenetics/tree/birthdeath/BirthDeathSamplingTreatmentProcess.cpp (hunk 4 above)
Comment (afmagee) [reply src/core/distributions/phylogenetics/tree/birthdeath/BirthDeathSamplingTreatmentProcess.cpp:617]: I didn't realize this was an issue, sure, let's keep it

File: src/core/distributions/phylogenetics/tree/birthdeath/BirthDeathSamplingTreatmentProcess.cpp (hunk 2 above)
Comment (bredelings) [reply src/core/distributions/phylogenetics/tree/birthdeath/BirthDeathSamplingTreatmentProcess.cpp:1619]: maybe add a note here about classification & tolerances being handled in countAllNodes / whichIntervalTime?

File: src/core/distributions/phylogenetics/tree/birthdeath/BirthDeathSamplingTreatmentProcess.cpp (hunk 4 above)
Comment (bredelings) [reply src/core/distributions/phylogenetics/tree/birthdeath/BirthDeathSamplingTreatmentProcess.cpp:617]: The patch is adding something... I guess @hoehna is asking why it was original removed?

**Review by hoehna — APPROVED** [review]
OK, looks good
